### PR TITLE
Feature/test enhancement

### PR DESCRIPTION
**Pull request by dhelmert**
### Reason for this PR
- Validation was neither explicitly tested, nor documented

### Changes in this PR
- Added documentation for validation
- Added tests for validation for Course-, Authentication-, User- and Matriculation-Management

### CODE REVIEW

File: product_code/authentication_service/impl/src/test/scala/de/upb/cs/uc4/authentication/impl/AuthenticationUserSpec.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      val errors = authUserValid.copy(username = "Ben").validate
+      val errorVariables = errors.map(error => error.name)
+      errorVariables should contain theSameElementsAs Seq("username")
+    }
+    "return a validation error for invalid character in username" in {
+      val errors = authUserValid.copy(username = "B€nn").validate
+      val errorVariables = errors.map(error => error.name)
+      errorVariables should contain theSameElementsAs Seq("username")
+    }
+    //PASSWORD
```
Comment (david-buderus) [inline product_code/authentication_service/impl/src/test/scala/de/upb/cs/uc4/authentication/impl/AuthenticationUserSpec.scala:29]: Here should be an empty line

File: product_code/authentication_service/impl/src/test/scala/de/upb/cs/uc4/authentication/impl/AuthenticationUserSpec.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package de.upb.cs.uc4.authentication.impl
```
Comment (david-buderus) [inline product_code/authentication_service/impl/src/test/scala/de/upb/cs/uc4/authentication/impl/AuthenticationUserSpec.scala:1]: This should be tested in the api package

File: product_code/course_service/impl/src/test/scala/de/upb/cs/uc4/course/impl/CourseSpec.scala [hunk 3]
```diff
@@ -0,0 +1,98 @@
+package de.upb.cs.uc4.course.impl
```
Comment (david-buderus) [inline product_code/course_service/impl/src/test/scala/de/upb/cs/uc4/course/impl/CourseSpec.scala:1]: This should be tested in the api package

File: product_code/matriculation_service/impl/src/test/scala/de/upb/cs/uc4/matriculation/impl/PutMessageMatriculationDataSpec.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package de.upb.cs.uc4.matriculation.impl
```
Comment (david-buderus) [inline product_code/matriculation_service/impl/src/test/scala/de/upb/cs/uc4/matriculation/impl/PutMessageMatriculationDataSpec.scala:1]: This should be tested in the api package

File: product_code/matriculation_service/impl/src/test/scala/de/upb/cs/uc4/matriculation/impl/PutMessageMatriculationDataSpec.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+class PutMessageMatriculationDataSpec extends AnyWordSpecLike with Matchers {
+  val putMessageValid: PutMessageMatriculationData = PutMessageMatriculationData("Computer Science", "SS2019")
+
+  "A PutMessageMatriculationData" should {
+    "be validated" in {
+      val errors = putMessageValid.validate
+      errors shouldBe empty
+    }
+    //FIELD OF STUDY
```
Comment (david-buderus) [inline product_code/matriculation_service/impl/src/test/scala/de/upb/cs/uc4/matriculation/impl/PutMessageMatriculationDataSpec.scala:15]: Here should be an empty line

File: product_code/matriculation_service/impl/src/test/scala/de/upb/cs/uc4/matriculation/impl/PutMessageMatriculationDataSpec.scala [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    "return a validation error for undefined fieldOfStudy" in {
+      val errors = putMessageValid.copy(fieldOfStudy = "ThisDoesNot Exist").validate
+      val errorVariables = errors.map(error => error.name)
+      errorVariables should contain theSameElementsAs Seq("fieldOfStudy")
+    }
+    //SEMESTER
```
Comment (david-buderus) [inline product_code/matriculation_service/impl/src/test/scala/de/upb/cs/uc4/matriculation/impl/PutMessageMatriculationDataSpec.scala:21]: Here should be an empty line

File: product_code/user_service/impl/src/test/scala/de/upb/cs/uc4/user/impl/UserSpec.scala [hunk 7]
```diff
@@ -7,62 +7,209 @@ import org.scalatest.wordspec.AnyWordSpecLike
...
-  "An Admin" should {
+  "A User" should {
     "be validated" in {
-      val errors = admin0.validate
+      val errors = adminValid.validate
       errors shouldBe empty
     }
-    "return validation errors" in {
-      val errors = admin1.validate
+    //USERNAME
```
Comment (david-buderus) [inline product_code/user_service/impl/src/test/scala/de/upb/cs/uc4/user/impl/UserSpec.scala:22]: Here should be an empty line

File: product_code/user_service/impl/src/test/scala/de/upb/cs/uc4/user/impl/UserSpec.scala [hunk 8]
```diff
@@ -7,62 +7,209 @@ import org.scalatest.wordspec.AnyWordSpecLike
```
Comment (david-buderus) [inline product_code/user_service/impl/src/test/scala/de/upb/cs/uc4/user/impl/UserSpec.scala:7]: This should be tested in the api package

File: product_code/user_service/impl/src/test/scala/de/upb/cs/uc4/user/impl/UserSpec.scala [hunk 9]
```diff
@@ -7,62 +7,209 @@ import org.scalatest.wordspec.AnyWordSpecLike
...
+    "return a validation error for incorrect length in matriculationId" in {
+      val errors = studentValid.copy(matriculationId = "00042").validate
       val errorVariables = errors.map(error => error.name)
-      errorVariables should contain only ("role", "email", "phoneNumber", "birthDate", "firstName", "lastName", "matriculationId", "latestImmatriculation")
+      errorVariables should contain theSameElementsAs Seq("matriculationId")
     }
   }
 
-  "An Adress" should {
+  "An Address" should {
```
Comment (david-buderus) [inline product_code/user_service/impl/src/test/scala/de/upb/cs/uc4/user/impl/UserSpec.scala:150]: Address could be checked in its own Spec (but then every UserType could be checked in its own Spec (maybe would be better)).

**Review by david-buderus — CHANGES_REQUESTED** [review]

File: product_code/user_service/impl/src/test/scala/de/upb/cs/uc4/user/impl/UserSpec.scala (hunk 9 above)
Comment (lrschaef) [reply product_code/user_service/impl/src/test/scala/de/upb/cs/uc4/user/impl/UserSpec.scala:150]: I agree, we COULD do that, but I dont really see the reason to do that now. We may do that if the file gets too large (which may very well be in this sprint, if we write more tests for users and the address)

File: product_code/authentication_service/api/src/test/scala/de/upb/cs/uc4/authentication/AuthenticationUserSpec.scala [hunk 10]
```diff
@@ -0,0 +1,38 @@
+package de.upb.cs.uc4.authentication
```
Comment (david-buderus) [inline product_code/authentication_service/api/src/test/scala/de/upb/cs/uc4/authentication/AuthenticationUserSpec.scala:1]: We could think about moving every test to the exact same folder as the corresponding class for better overview.
(But not part of this PR).

**Review by david-buderus — APPROVED** [review]

**Review by ldklenner — APPROVED** [review]

**Review by lrschaef — APPROVED** [review]
